File: RachitGoyal/InsertionSort.cpp
#include<bits/stdc++.h>
using namespace std;
int main()
{
    int n;
    cin>>n;
    int arr[n];
    for(int i=0;i<n;i++)
    {
        cin>>arr[i];
    }
    int temp;
    for(int i=1;i<n;i++)
    {
        int k=arr[i];
        for(int j=i-1;j>=0;j--)
        {
    
            if(k<arr[j])
            {
                temp=arr[j];
                arr[j]=k;
                arr[j+1]=temp;
            }
        }
    }
    for(int i=0;i<n;i++)
    {
        cout<<arr[i]<<' ';
    }
    
    return 0;
}


//Space Complexity: O(1)
//Time Complexity:-
//Worst Case: O(n^2)
//Best Case:O(n)

File: Prashant/Sorting/InsertionSort.java

/** this program shows the implementation of insertion sort */
import java.io.IOException;
import java.util.Scanner;

public class InsertionSort {
    public static void insertionSort(int array[]) {
        int n = array.length;
        for (int j = 1; j < n; j++) {
            int key = array[j];
            int i = j - 1;
            while ((i > -1) && (array[i] > key)) {
                array[i + 1] = array[i];
                i--;
            }
            array[i + 1] = key;
        }
    }

    public static void main(String a[]) {
        int[] arr1 = { 9, 14, 3, 2, 43, 11, 58, 22 };
        System.out.println("Before Insertion Sort");
        for (int i : arr1) {
            System.out.print(i + " ");
        }
        System.out.println();

        insertionSort(arr1);// sorting array using insertion sort

        System.out.println("After Insertion Sort");
        for (int i : arr1) {
            System.out.print(i + " ");
        }
    }
}

// Worst Case Time Complexity [ Big-O ]: O(n2)

// Best Case Time Complexity [Big-omega]: O(n)

// Average Time Complexity [Big-theta]: O(n2)

// Space Complexity: O(1)
